server/src/test/java/io/crate/execution/IncrementalPageBucketReceiverTest.java

```diff
@@ -54,4 +65,75 @@ public void test_processing_future_completed_when_finisher_throws() {
+        TestingRowConsumer batchConsumer = new TestingRowConsumer();
+        Collector<Row, Object, Iterable<Row>> collector = new Collector<>() {
+            @Override
+            public Supplier<Object> supplier() {
+                return ArrayList::new;
+            }
+
+            @Override
+            public BiConsumer<Object, Row> accumulator() {
+                return (_, _) -> {
+                    throw new CircuitBreakingException("dummy");
+                };
+            }
+
+            @Override
+            public BinaryOperator<Object> combiner() {
+                return null;
+            }
+
+            @Override
+            public Function<Object, Iterable<Row>> finisher() {
+                return null;
+            }
+
+            @Override
+            public Set<Characteristics> characteristics() {
+                return Set.of();
+            }
+        };
+
+        var pageBucketReceiver = new IncrementalPageBucketReceiver<>(
+            collector,
+            batchConsumer,
+            Runnable::run,
+            new Streamer[1],
+            1
+        );
+
+        // First call goes to currentlyAccumulating == null, needMore must be true after the call
+        final CompletableFuture<DistributedResultResponse> result = new CompletableFuture<>();
+        PageResultListener listener = needMore -> result.complete(new DistributedResultResponse(needMore));
+        pageBucketReceiver.setBucket(0, Bucket.EMPTY, false, listener);
+        assertThat(result).isCompletedWithValueMatchingWithin(
+            distributedResultResponse -> distributedResultResponse.needMore() == true,
+            Duration.ofSeconds(1)
+        );
+
+        // Second call goes to currentlyAccumulating != null, use non-empty bucket to provoke CBE
+        Bucket bucket = new ArrayBucket(new Object[][]{
+            new Object[]{1},
+        });
+        final CompletableFuture<DistributedResultResponse> result2 = new CompletableFuture<>();
+        PageResultListener listener2 = needMore -> result2.complete(new DistributedResultResponse(needMore));
+        pageBucketReceiver.setBucket(0, bucket, false, listener2);
+        assertThat(result2).isCompletedWithValueMatchingWithin(
+            distributedResultResponse -> distributedResultResponse.needMore() == true,
+            Duration.ofSeconds(1)
+        );
+
+        // Call after failed processRows, listener must see that previous call was completed exceptionally
+        final CompletableFuture<DistributedResultResponse> result3 = new CompletableFuture<>();
+        PageResultListener listener3 = needMore -> result3.complete(new DistributedResultResponse(needMore));
+        pageBucketReceiver.setBucket(0, Bucket.EMPTY, false, listener3);
+        assertThat(result3).isCompletedWithValueMatchingWithin(
+            distributedResultResponse -> distributedResultResponse.needMore() == false,
+            Duration.ofSeconds(1)
```

Comment:
call with last = false coming right after  
the call when `processRows` threw used to skip an exception and needMore was wrongly true. If I remove try/catch inside whenComplete, this assertion fails